File: 648-replace-words/replace-words.cpp
class Solution {
public:
    string replaceWords(vector<string>& dictionary, string sentence) {
        string token; 
        stringstream ss(sentence);
        vector<string> s;
        while (getline(ss, token, ' ')) {
            s.push_back(token); 
        }
        unordered_set<string> root(dictionary.begin(), dictionary.end()); 
        string res = ""; 

        for (int i = 0; i < s.size(); ++i) {
            string word = s[i]; 
            string builder = "";
            bool hasRoot = false; 
            for (char c : word) {
                builder += string(1, c); 
                if (root.find(builder) != root.end()) {
                    res += builder;
                    hasRoot = true; 
                    break; 
                }
            }
            if (!hasRoot) {
                res += builder; 
            }

            if (i < s.size() - 1) {
                res += " "; 
            }
        }

        return res; 
    }
};
/*
brute force approach iterate each word in the sentence
    the moment we find the first matching (also the shortest matchign)
    replace it
*/ 

File: 648-replace-words/replace-words.java
class Solution {
    public String replaceWords(List<String> dictionary, String sentence) {
        String[] wordArray = sentence.split(" "); 
        Set<String> dictSet = new HashSet<>(dictionary); 
        
        for (int i = 0; i < wordArray.length; ++i) {
            wordArray[i] = shortestRoot(wordArray[i], dictSet); 
        }

        return String.join(" ", wordArray); 
    }

    public String shortestRoot(String word, Set<String> dict) {
        StringBuilder temp = new StringBuilder(); 
        for (int i = 0; i < word.length(); ++i) {
            temp.append(word.charAt(i));
            if (dict.contains(temp.toString())) {
                return temp.toString();
            }
        }
        return temp.toString(); 
    }
}